SIPS/sip-7.md

```diff
@@ -2,307 +2,205 @@
-};
-```
+An interface created by a snap SHALL be declarative in form of a JSON-serializable JavaScript value.
 
-### Initial components
+## Base Components
 
-This SIP describes a couple of components for the initial version of the UI:
+All component types inside UI declaration SHALL be derived from either `Parent` or `Literal`.
 
-1. A panel component, which renders a list of `children` components.
-2. A heading component, rendering the text as a HTML heading (`h1`, `h2`, `h3`, and so on).
-    1. The heading is rendered dynamically based on the nesting of panels, i.e.:
-        1. A heading in a panel at the root SHOULD render a `h1`.
-        2. A heading in a panel nested in the root panel SHOULD render a `h2`.
-        3. etc.
-3. A plain text component, rendering the text as a HTML paragraph (`p`).
-4. A button component, rendering the text as a HTML button (`button`), with an optional ID for handling user input.
-    1. The button can have two variants: `primary` and `secondary`.
-    2. If the variant is not specified, `primary` SHOULD be used.
-5. A divider component, rendering a HTML thematic break (`hr`).
-6. A spacer component, pushing content around it to the top and bottom, similar to how `flex-grow` works.
-7. A spinner component, rendering a spinner to indicate loading data.
+### Node
 
-```tsx
-type ComponentType = 'panel' | 'heading' | 'text' | 'button' | 'divider' | 'spacer' | 'spinner';
+The base type for all the components.
 
-interface PanelComponent extends BaseComponent {
-  type: 'panel';
-  children: Component[];
+```typescript
+interface Node {
+  type: string;
 }
+```
 
-interface HeadingComponent extends BaseComponent {
-  type: 'heading';
-  text: string;
-}
+### Parent
 
-interface TextComponent extends BaseComponent {
-  type: 'text';
-  text: string;
-}
-
-interface ButtonComponent extends BaseComponent {
-  type: 'button';
-  variant?: 'primary' | 'secondary';
-  text: string;
-  name?: string;
+```typescript
+interface Parent extends Node {
+  children: Node[];
 }
+```
 
-interface DividerComponent extends BaseComponent {
-  type: 'divider';
-}
+### Literal
 
-interface SpacerComponent extends BaseComponent {
-  type: 'spacer';
+```typescript
+interface Literal extends Node {
+  value: any;
 }
-
-interface SpinnerComponent extends BaseComponent {
-  type: 'spinner';
-}
-
-type Component =
-  | PanelComponent
-  | HeadingComponent
-  | TextComponent
-  | ButtonComponent
-  | DividerComponent
-  | SpacerComponent
-  | SpinnerComponent;
 ```
 
-### JSON-RPC methods
-
-#### `snap_showInterface`
-
-The `snap_showInterface` method lets Snaps show a user interface arbitrarily. The params MUST be an object, with the following properties:
+## Core Components
 
-- `ui` (`Component`): The custom UI to show.
+> Defining a complete set of components used in the wallet is outside of scope for this SIP.
+>
+> Only a core set of components is defined, delegating definition of a complete set of useful components to other SIPs.
```

Comment:
Why? We had already defined a good set of initial components, other SIPs can expand on this.

Reply:
I'll write a second SIP in this PR as well.

This will allow this SIP to become Final, specifying the API in general, and the second to be considered Living, and be continually updated with new components as we go.